### PR TITLE
Adding more test cases for testing multiple http methods

### PR DESCRIPTION
**Pull request by sebing**
Fixing the issue #458

Adding more tests for TkProxyTest

### CODE REVIEW

**Comment by davvd** [comment]
@sebing I will find a reviewer for your pull requests shortly, thanks for contribution!

**Comment by davvd** [comment]
@dmzaytsev review this one, please

File: src/test/java/org/takes/tk/TkProxyTest.java [hunk 1]
```diff
@@ -34,34 +34,69 @@
...
- *  The tests should verify the different HTTP methods (GET, POST, etc),
- *  as well as the different combinations of request/response headers.
- */
+  */
 public final class TkProxyTest {
 
     /**
-     * TkProxy can work.
-     * @throws Exception If some problem inside
+     * TkProxy Testcase for testing the creation of the object with method POST.
```
Comment (dmzaytsev) [inline src/test/java/org/takes/tk/TkProxyTest.java]: @sebing please update java doc according to this article www.yegor256.com/2014/04/27/typical-mistakes-in-java-code.html?expand=1#test-method-names

same below

**Comment by dmzaytsev** [comment]
@sebing see my comment above

**Comment by sebing** [comment]
@dmzaytsev  thanks for comments. I corrected them. 
And thanks for the link, it had more guidelines(that clear lot of doubts). I like them and follow them.

**Comment by dmzaytsev** [comment]
@sebing thank you

**Comment by dmzaytsev** [comment]
@rultor merge

File: src/test/java/org/takes/tk/TkProxyTest.java [hunk 2]
```diff
@@ -34,34 +34,69 @@
...
+    public void correctlyMapsPathStringOnPost() throws Exception {
+        this.correctlyMapsPathString(RqMethod.POST);
+    }
+
+    /**
+     * A private method to do test with multiple httpMethods.
+     * @param httpmethod HTTP methods (GET, POST, etc),
+     * @throws Exception If some problem inside.
+     */
+    private void justWorks(final String httpmethod) throws Exception {
```
Comment (yegor256) [inline src/test/java/org/takes/tk/TkProxyTest.java]: @sebing why not just `method`? see http://www.yegor256.com/2015/01/12/compound-name-is-code-smell.html

File: src/test/java/org/takes/tk/TkProxyTest.java [hunk 3]
```diff
@@ -71,11 +106,13 @@ public void exec(final URI home) throws IOException {
...
-     * TkProxy can work.
-     * @throws Exception If some problem inside
+     * A private method to call the TkProxy. act with httpMethod.
+     *
+     * @param httpmethod HTTP methods (GET, POST, etc),
+     * @throws Exception If some problem inside.
      */
-    @Test
-    public void correctlyMapsPathString() throws Exception {
+    private void correctlyMapsPathString(final String httpmethod)
```
Comment (yegor256) [inline src/test/java/org/takes/tk/TkProxyTest.java]: @sebing same here, what's wrong about `method`?

**Comment by yegor256** [comment]
@sebing a few comments, see above

**Comment by sebing** [comment]
@yegor256 Fixed the stingy compound names.

**Comment by sebing** [comment]
@yegor256 @davvd Any update on the closure of this merge?

**Comment by yegor256** [comment]
@sebing I don't like this design of test methods. I should probably write an article about it, but here is what I mean, in short. This is your code:

```
@Test
void doesThis() {
  this.works(1);
}
@Test
void doesThat() {
  this.works(2);
}
```

This is how it should be instead:

```
@Test
void doesThis() {
  assertThat(this.asset(1), is(true));
}
@Test
void doesThat() {
  assertThat(this.asset(2), is(true));
}
```

My point is that you should **not** create a test method that doesn't have an assertion. By doing that you completely destroys the idea of a unit test. The test becomes unclear and unmaintainable.

**Comment by sebing** [comment]
@yegor256 I agree with you.
But I thought this is doing the same.

```java
MatcherAssert.assertThat(
                        new RsPrint(
                            new TkProxy(home).act(new RqFake(method))
                        ).print(),
                        Matchers.containsString("hello")
```
@dmzaytsev
To make the code more redable,
this assertion is happening in a private method, should i move it to the test case? what do you think of this?

**Comment by sebing** [comment]
@dmzaytsev  what do you think? How can we improve the code?

**Comment by sebing** [comment]
@yegor256 @dmzaytsev  I know, this is not a school :) still...
- What is the difference between assertThat in the test  and in a private method?
- Is it bad to assert in a private method? ( i cannot find a clear best practice in google)

**Comment by sebing** [comment]
@yegor256 If the code with assert in the private methods, can you accept the code?

**Comment by dmzaytsev** [comment]
@sebing just do it :)

https://github.com/yegor256/takes/pull/551#issuecomment-178987626

> yegor256 commented 2 days ago
> My point is that you should not create a test method that doesn't have an assertion. By doing that you completely destroys the idea of a unit test. The test becomes unclear and unmaintainable.

**Comment by sebing** [comment]
@dmzaytsev Can you re-review the changes?

**Comment by dmzaytsev** [comment]
@sebing If I had to do that task, I would suggest something like this
```
final String [] methods = {RqMethod.POST, RqMethod.GET, .....};
final String content = "content";
for(final String method: methods) {
    new FtRemote(new TkFixed(content)).exec(
            new FtRemote.Script() {
                @Override
                public void exec(final URI home) throws IOException {
                    MatcherAssert.assertThat(
                       String.format("method %s", method),
                        new RsPrint(
                            new TkProxy(home).act(new RqFake(method))
                        ).print(),
                        Matchers.containsString(content)
                    );
                }
            }
        );
}
```

**Comment by sebing** [comment]
@dmzaytsev Is there any other way we can test?
I am getting this AppVeyor check error:
```
Avoid instantiating new objects inside loops
```

**Comment by dmzaytsev** [comment]
@sebing `@SuppressWarnings("PMD.AvoidInstantiatingObjectsInLoops")`

**Comment by sebing** [comment]
@dmzaytsev It is a better implementation with the loop. 
May be i was deceived by the task name(but no excuse). Can you review the test again?

File: src/test/java/org/takes/tk/TkProxyTest.java [hunk 4]
```diff
@@ -50,30 +51,43 @@
 public final class TkProxyTest {
 
     /**
-     * TkProxy can work.
+     * An array of http methods for testing.
+     */
+    private static final String[] METHODS = {
```
Comment (dmzaytsev) [inline src/test/java/org/takes/tk/TkProxyTest.java]: @sebing what about other http methods ?

**Comment by dmzaytsev** [comment]
@sebing just one comment above

File: src/test/java/org/takes/tk/TkProxyTest.java (hunk 4 above)
Comment (sebing) [inline src/test/java/org/takes/tk/TkProxyTest.java]: @dmzaytsev The requirement was not clear for me :) 
I added the methods except( CONNECT and HEAD, they need a different testing).
Should I add the testcase for them too?

Comment (sebing) [inline src/test/java/org/takes/tk/TkProxyTest.java]: @dmzaytsev  I will add another test case for HEAD
And regarding the Connect
```
java.io.IOException: Failed CONNECT request to http://localhost:52991/

	at com.jcabi.http.request.JdkRequest$1.send(JdkRequest.java:127)
	at com.jcabi.http.request.BaseRequest.fetchResponse(BaseRequest.java:336)
	at com.jcabi.http.request.BaseRequest.fetch(BaseRequest.java:251)
	at org.takes.tk.TkProxy.act(TkProxy.java:119)
	at org.takes.tk.TkProxyTest$2.exec(TkProxyTest.java:94)
	at org.takes.http.FtRemote.exec(FtRemote.java:157)
	at org.takes.tk.TkProxyTest.justWorksOnConnect(TkProxyTest.java:90)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:606)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
	at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:69)
	at com.intellij.rt.execution.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:234)
	at com.intellij.rt.execution.junit.JUnitStarter.main(JUnitStarter.java:74)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:606)
	at com.intellij.rt.execution.application.AppMain.main(AppMain.java:144)
Caused by: java.net.ProtocolException: Invalid HTTP method: CONNECT
	at java.net.HttpURLConnection.setRequestMethod(HttpURLConnection.java:428)
	at com.jcabi.http.request.JdkRequest$1.send(JdkRequest.java:102)
	... 32 more
```
Should I add a puzzle?

File: src/test/java/org/takes/tk/TkProxyTest.java [hunk 5]
```diff
@@ -46,24 +47,70 @@
  * @todo #377:30min/DEV We need more tests for TkProxy.
```
Comment (dmzaytsev) [inline src/test/java/org/takes/tk/TkProxyTest.java]: @sebing you have to delete the original puzzle

File: src/test/java/org/takes/tk/TkProxyTest.java [hunk 6]
```diff
@@ -46,24 +47,70 @@
...
  *  The tests should verify the different HTTP methods (GET, POST, etc),
  *  as well as the different combinations of request/response headers.
+ *
+ * @todo #458:30min We need to support the http method CONNECT.
```
Comment (dmzaytsev) [inline src/test/java/org/takes/tk/TkProxyTest.java]: @sebing I think we don't need to proxy `CONNECT` requests

File: src/test/java/org/takes/tk/TkProxyTest.java [hunk 7]
```diff
@@ -46,24 +47,70 @@
...
+        RqMethod.GET,
+        RqMethod.PUT,
+        RqMethod.DELETE,
+        RqMethod.TRACE,
+    };
+
+    /**
+     * The expected result for the Head http request.
+     */
+    private static final String RESULT = "";
```
Comment (dmzaytsev) [inline src/test/java/org/takes/tk/TkProxyTest.java]: @sebing why you need this variable?

File: src/test/java/org/takes/tk/TkProxyTest.java [hunk 8]
```diff
@@ -46,24 +47,70 @@
...
                 public void exec(final URI home) throws IOException {
                     MatcherAssert.assertThat(
                         new RsPrint(
-                            new TkProxy(home).act(new RqFake("PUT"))
+                            new TkProxy(home).act(
+                                new RqFake(RqMethod.HEAD)
+                            )
                         ).print(),
-                        Matchers.containsString("hello")
+                        Matchers.containsString(TkProxyTest.RESULT)
```
Comment (dmzaytsev) [inline src/test/java/org/takes/tk/TkProxyTest.java]: @sebing I didn't get what you check here ?
I guess any string contains `""`, isn't it?

Anyway I think the existing methods enough for the test, let's remove test for  `HEAD`

**Comment by dmzaytsev** [comment]
@sebing please see my comments

**Comment by sebing** [comment]
@dmzaytsev I corrected the suggestions. 
- Removed the HEAD and Connect methods.
- Removed the puzzle.
You have any idea one of the travis-ci(java 8) job is failing?

**Comment by dmzaytsev** [comment]
@sebing try close/open this PR to trigger a new Travis build

**Comment by sebing** [comment]
@dmzaytsev trying to re run the travis

**Comment by dmzaytsev** [comment]
@sebing feel free to submit a bug about the unstable test on java 8

**Comment by sebing** [comment]
@dmzaytsev Now the build is success. Can you review them?

**Comment by dmzaytsev** [comment]
@sebing looks good
thanks

**Comment by dmzaytsev** [comment]
@rultor merge

**Comment by sebing** [comment]
@yegor256 Any update on this PR?

**Comment by sebing** [comment]
@yegor256 Can you look at this PR?

**Comment by yegor256** [comment]
@rultor try to merge

**Comment by sebing** [comment]
@yegor256  can we try to merge this again?

**Comment by sebing** [comment]
@yegor256 The merge failed, can we try to merge this again?

**Comment by sebing** [comment]
@yegor256 @davvd The rultor merge failed, can we try again?

**Comment by sebing** [comment]
@yegor256 @dmzaytsev @davvd What can we do here? should we try to ask rultor to merge again?

**Comment by dmzaytsev** [comment]
@sebing only @yegor256 can confirm the merge request as hi is the architect on this project
we should wait until Yegor return from a business trip

**Comment by sebing** [comment]
@dmzaytsev Thanks for the information. 
@davvd is the information that the architect is not available published anywhere? So I won't be keep spamming people without this information next time? :)

**Comment by yegor256** [comment]
@rultor try to merge again

**Comment by davvd** [comment]
@elenavolokhova please, help us to evaluate quality here, as per [par.24](http://at.teamed.io/policy.html#24)

**Comment by davvd** [comment]
@rultor deploy pls

**Comment by elenavolokhova** [comment]
@davvd Looks good!

**Comment by davvd** [comment]
> @davvd Looks good!

@elenavolokhova thanks for the review

**Comment by davvd** [comment]
@dmzaytsev I added 10 mins to @elenavolokhova (for QA review) in transaction 78630815; thanks, I just added **34 mins** to your account, payment `78630822`, 460 hours and 9 mins spent; review comments (c=19) added as a bonus; +34 added to your rating, at the moment it is: [+3514](http://www.netbout.com/b/35166?open=rating)

**Comment by sebing** [comment]
@dmzaytsev Can you help me understand this. This PR is merged and deployed. But the puzzle is still in the master. Any clue why?

**Comment by dmzaytsev** [comment]
@sebing no idea
but I would suggest you create a new PR and remove the puzzle

**Comment by sebing** [comment]
@dmzaytsev it is not the puzzle alone, the whole changes are not merged. So should I redo the work?

**Comment by dmzaytsev** [comment]
@yegor256 take a look please

**Comment by original-brownbear** [comment]
@sebing @dmzaytsev this is not Rultor. Please look at 79be7637a3e6eed12c0f666988c0fc8151e20482
The last commit to t his branch was simply an apparently wrong merge with upstream reverting a bunch of the prior changes.
I mean Rultor is not setting the "merged" status anyhow, this is GitHub. Unless you see a commit by Rultor, this cannot be Rultor's fault.

Makes sense?

**Comment by sebing** [comment]
@dmzaytsev @original-brownbear So how do we proceed now?

**Comment by original-brownbear** [comment]
@sebing you need to create a new PR and ask for another review. Obviously this one was wrong and got merged accidentally.
I suggest simply going to the commit before your 79be763 and correctly rebasing that onto `master`.
For all other questions on this your architect here, @yegor256 is who you should though ;)

**Comment by yegor256** [comment]
@sebing looks like it was your mistake, as an author of PR. just do it again.

**Pull request merged**